source/credentials_provider_sts_web_identity.c

```diff
@@ -0,0 +1,1120 @@
+    struct aws_allocator *allocator) {
+
+    struct aws_credentials_provider_sts_web_identity_parameters *parameter =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_credentials_provider_sts_web_identity_parameters));
+    if (parameter == NULL) {
+        return NULL;
+    }
+    parameter->allocator = allocator;
+
+    bool success = false;
+    struct aws_string *region = NULL;
+    struct aws_string *role_arn = NULL;
+    struct aws_string *role_session_name = NULL;
+    struct aws_string *token_file_path = NULL;
+
+    /* check environment variables */
+    aws_get_environment_value(allocator, s_sts_web_identity_region_env, &region);
+    aws_get_environment_value(allocator, s_sts_web_identity_role_arn_env, &role_arn);
+    aws_get_environment_value(allocator, s_sts_web_identity_role_session_name_env, &role_session_name);
+    aws_get_environment_value(allocator, s_sts_web_identity_token_file_path_env, &token_file_path);
+
+    /**
+     * check config profile if either role_arn or token_file_path or role_session_name is not resolved from environment
+     * variable. Region is not enforced, but we need it to construct sts endpoint, if we can't find region from
+     * environment, we should check if it's set in config file, if neither, use default value "us-east-1" as region.
+     * Role session name can be generated by us using uuid if not found.
+     */
+    struct aws_profile_collection *config_profile = NULL;
+    struct aws_string *profile_name = NULL;
+    struct aws_profile *profile = NULL;
+    bool good_region = (region && region->len);
+    bool good_role_arn_and_token_file_path = (role_arn && role_arn->len && token_file_path && token_file_path->len);
+    if (!good_region || !good_role_arn_and_token_file_path) {
+        config_profile = s_sts_web_identity_load_profile(allocator);
+        profile_name = aws_get_profile_name(allocator, &s_default_profile_name_cursor);
+        if (config_profile && profile_name) {
+            profile = aws_profile_collection_get_profile(config_profile, profile_name);
+        }
+
+        if (!config_profile || !profile_name || !profile) {
+            /* if region is also not in good shape, we just need to use default us-east-1. */
+            if (!good_role_arn_and_token_file_path) {
+                AWS_LOGF_ERROR(
+                    AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                    "Failed to resolve role arn or token file path during sts web identity provider initialization.")
+                goto on_finish;
+            }
+
+        } else {
+            if (!region || !region->len) {
+                struct aws_profile_property *region_property =
+                    aws_profile_get_property(profile, s_sts_web_identity_region_config);
+                if (region_property) {
+                    region = aws_string_new_from_string(allocator, region_property->value);
+                }
+            }
+
```

Comment:
These 4 if-blocks should get relegated to a helper function.  Also the strings can be unconditionally destroyed.